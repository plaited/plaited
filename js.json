{
  "compilerOptions": {
    "jsx": "react-jsx",
    "jsxImportSource": "plaited",
    "paths": {
      "plaited/jsx-runtime": [
        "./src/jsx/runtime.ts"
      ],
      "plaited/jsx-dev-runtime": [
        "./src/jsx/dev-runtime.ts"
      ]
    }
  }
}