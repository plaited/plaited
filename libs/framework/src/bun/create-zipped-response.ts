export const jsMimeTypes = ['text/javascript;charset=utf-8', 'application/json;charset=utf-8'] as const

export const isJsMimeType = (type: string): type is (typeof jsMimeTypes)[number] =>
  jsMimeTypes.includes(type as (typeof jsMimeTypes)[number])

export const createZippedResponse = (str: string, type: (typeof jsMimeTypes)[number]) => {
  const compressed = Bun.gzipSync(Buffer.from(str))
  return new Response(compressed, {
    headers: {
      'content-type': type,
      'content-encoding': 'gzip',
    },
  })
}
