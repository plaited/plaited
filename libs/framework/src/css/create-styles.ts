import type { CSSProperties, CSSPropertiesObjectLiteral, CSSClasses, CreateStylesObjects } from './types.js'
import { hashString, kebabCase } from '../utils.js'

const createClassHash = (...args: (string | number)[]) =>
  hashString(args.join(' '))?.toString(36).replace(/^-/g, '_') ?? ''

const isPrimitive = (val: string | number | CSSPropertiesObjectLiteral<string>): val is string | number =>
  typeof val === 'string' || typeof val === 'number'

const formatStyles = ({
  map,
  value,
  prop,
  selectors = [],
}: {
  map: Map<string, string>
  value: CSSPropertiesObjectLiteral<typeof prop> | CSSProperties[typeof prop]
  prop: string
  selectors?: string[]
}) => {
  if (!value) return
  if (isPrimitive(value)) {
    const hash = createClassHash(prop, value, ...selectors)
    const className = `p${hash}`
    const rule = `${kebabCase(prop)}:${value};`
    if (!selectors.length) return map.set(className, `.${className}{${rule}}`)
    const arr = selectors.map((str) => (str.startsWith('@') ? `${str}{` : `&${str}{`))
    return map.set(className, `.${className}{${arr.join('')}${rule}${'}'.repeat(arr.length)}}`)
  }
  const arr = Object.entries(value)
  const length = arr.length
  for (let i = 0; i < length; i++) {
    const [context, val] = arr[i]
    if (context === 'default' || /^(:|\[|(@(container|layer|media|supports)))/.test(context)) {
      const nextSelectors = [...selectors]
      context !== 'default' && nextSelectors.push(context)
      formatStyles({ map, value: val, prop, selectors: nextSelectors })
    }
  }
}

/** A types safe function for creating hashed utility className(s) and stylesheet(s) */
export const createStyles = <T extends CSSClasses>(classNames: T) =>
  Object.entries(classNames).reduce((acc, [cls, props]) => {
    const map = new Map<string, string>()
    for (const prop in props) formatStyles({ map, prop, value: props[prop] })
    const classes = [...map.keys()]
    const hash = createClassHash(...classes)
    acc[cls as keyof T] = {
      className: [`${cls}_${hash}`, ...classes].join(' '),
      stylesheet: [...map.values()],
    }
    return acc
  }, {} as CreateStylesObjects<T>)
