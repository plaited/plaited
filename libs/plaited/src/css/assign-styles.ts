import { AssignStylesObject } from '../types.js'

export const assignStyles = (...styleObjects: Array<AssignStylesObject | undefined | false | null>) => {
  const cls: Array<string | undefined | false | null> = []
  const style: Array<string | undefined | false | null> = []
  for (const styleObject of styleObjects) {
    if (!styleObject) continue
    const { className, stylesheet } = styleObject
    className && cls.push(...(Array.isArray(className) ? className : [className]))
    stylesheet && style.push(...(Array.isArray(stylesheet) ? stylesheet : [stylesheet]))
  }
  return { className: cls, stylesheet: style }
}
