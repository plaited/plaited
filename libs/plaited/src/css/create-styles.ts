import { hashString, kebabCase } from '@plaited/utils'
import { CSSProperties, CSSPropertiesObjectLiteral, CSSClasses, StyleObjects } from '../types.js'

const createClassHash = (...args: (string | number)[]) => 'p' + hashString(args.join(' '))?.toString(36)

const isPrimitive = (val: string | number | CSSPropertiesObjectLiteral<string>): val is string | number =>
  typeof val === 'string' || typeof val === 'number'

const handleAtRules = ({
  prop,
  val,
  hash,
  selector,
  atRules,
  last,
}: {
  prop: string
  val: string | number
  hash: string
  selector: string
  atRules: string[]
  last: number
}) =>
  atRules.reduceRight<string>((acc, cur, i) => {
    if (i === last) return `${cur}{.${hash}${selector}{${kebabCase(prop)}:${val}}}`
    return `${cur}{${acc}}`
  }, '')

const formatStyles = <T extends CSSClasses>({
  cls,
  stylesheets,
  value,
  prop,
  selectors = [],
}: {
  cls: string
  stylesheets: Map<keyof StyleObjects<T>, Map<string, string>>
  value: CSSPropertiesObjectLiteral<typeof prop> | CSSProperties[typeof prop]
  prop: string
  selectors?: string[]
}) => {
  const val = isPrimitive(value) ? value : value.default
  const map = stylesheets.get(cls) ?? stylesheets.set(cls, new Map<string, string>()).get(cls)!
  const [selector = '', ...atRules] = selectors
  if (val) {
    const hash = createClassHash(val, prop, selector, ...atRules)
    if (map.has(hash)) return
    const length = atRules.length
    map.set(
      hash,
      length ?
        handleAtRules({
          prop,
          val,
          hash,
          selector,
          atRules,
          last: length - 1,
        })
      : `.${hash}${selector}{${kebabCase(prop)}:${val}}`,
    )
  }
  if (!isPrimitive(value)) {
    const { default: _, ...rest } = value
    for (const context in rest) {
      formatStyles<T>({
        cls,
        prop,
        stylesheets,
        value: rest[context as keyof typeof rest],
        selectors: [
          context.startsWith(':') ? selector + context : selector,
          ...(/^@(container|layer|media|supports)/.test(context) ? [...atRules, context] : atRules),
        ],
      })
    }
  }
}

/** A types safe function for creating hashed utility className(s) and stylesheet(s) */
export const createStyles = <T extends CSSClasses>(
  classNames: T,
): StyleObjects<T> & {
  $: (key: string) => string | undefined
} => {
  const stylesheets = new Map<keyof StyleObjects<T>, Map<string, string>>()
  for (const cls in classNames) {
    const props = classNames[cls]
    for (const prop in props) {
      const value = props[prop]
      if (isPrimitive(value)) {
        const map = stylesheets.get(cls) ?? stylesheets.set(cls, new Map<string, string>()).get(cls)!
        const hash = createClassHash(value, prop, '')
        const sheet = `.${hash}{${kebabCase(prop)}:${value}}`
        map.set(`${hash}`, sheet)
      } else {
        formatStyles<T>({
          cls,
          prop,
          stylesheets,
          value,
        })
      }
    }
  }
  const toRet: StyleObjects<T> = {} as StyleObjects<T>
  for (const [name, sheets] of stylesheets) {
    const classNames = [...sheets.keys()].join(' ')
    toRet[name] = {
      className: `${name as string}_${hashString(classNames)?.toString(36) ?? ''} ${classNames}`,
      stylesheet: [...sheets.values()],
    }
  }
  return {
    ...toRet,
    $(key: string) {
      const obj = toRet[key]
      if (obj) return obj.className.split(' ')[0]
    },
  }
}
