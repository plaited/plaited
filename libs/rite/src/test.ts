/// <reference lib="dom" />
/// <reference lib="dom.iterable" />

import { TestResultError, TestResult, BrowserSessionResult } from '@web/test-runner-core/browser/session.js'
import { assert, Assertion, AssertionError } from './assert.js'

type TestCallback = (t: Assertion) => Promise<void> | void

interface Test {
  (name: string, cb: TestCallback): void
  skip: (name: string, cb: TestCallback) => void
}

type TestCase = () => Promise<boolean> | boolean

class TestRunner {
  #completedTests = 0
  #failedTests = 0
  #errors: TestResultError[] = []
  #cases: TestCase[] = []
  #timeout = 5_000
  #tests: TestResult[] = []
  updateTimeout(time: number) {
    this.#timeout = time
  }
  get timeout() {
    return this.#timeout
  }
  addCase(test: TestCase) {
    this.#cases.push(test)
  }
  addResult(result: TestResult) {
    this.#tests.push(result)
  }
  addError(obj: TestResultError) {
    this.#errors.push(obj)
  }
  async run(name: string) {
    for (const testCase of this.#cases) {
      try {
        const passed = await testCase()
        this.#completedTests++
        if (!passed) {
          this.#failedTests++
        }
      } catch (error) {
        console.error(error)
      }
    }

    if (this.#completedTests === this.#cases.length) {
      const results: BrowserSessionResult = {
        passed: !this.#failedTests,
        testResults: {
          name,
          suites: [],
          tests: this.#tests,
        },
        errors: this.#errors,
      }
      return results
    } else {
      throw new AssertionError(`Not all test completed`)
    }
  }
}

declare global {
  interface Window {
    __rite_test_runner?: TestRunner
  }
}

window.__rite_test_runner = new TestRunner()

const throwTimeoutError = async () => {
  const timeout = window.__rite_test_runner.timeout
  await assert.wait(timeout)
  throw new Error(`test takes longer than ${timeout}ms to complete`)
}

export const test: Test = (name, callback) => {
  window.__rite_test_runner.addCase(async () => {
    const startTime = performance.now()
    const result: TestResult = {
      name,
      passed: true,
      skipped: false,
    }
    try {
      await Promise.race([callback(assert), throwTimeoutError()])
      const msg = `✓ ${name}`
      // eslint-disable-next-line no-console
      console.log(msg)
    } catch (error) {
      const msg = `✗ ${name}`
      // eslint-disable-next-line no-console
      console.log(msg)
      result.passed = false
      if (error instanceof AssertionError) {
        const obj: Pick<TestResultError, 'actual' | 'expected' | 'message'> = JSON.parse(error.message)
        result.error = { name, stack: `${error.stack}`, ...obj }
        window.__rite_test_runner.addError({ name, ...obj })
      } else {
        throw new Error(error)
      }
    }
    result.duration = performance.now() - startTime
    window.__rite_test_runner.addResult(result)
    return result.passed
  })
}
test.skip = (name: string, _: TestCallback) => {
  window.__rite_test_runner.addCase(() => {
    const result = {
      name,
      passed: true,
      skipped: true,
    }
    window.__rite_test_runner.addResult(result)
    return true
  })
}
