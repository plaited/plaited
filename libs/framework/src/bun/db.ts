import { Database } from "bun:sqlite";

export const db = new Database(":memory:");
db.query(`create table templates;`);