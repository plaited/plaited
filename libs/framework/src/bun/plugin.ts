import type {  BunPlugin } from "bun";
import { db } from "./db.js";

export const plugin: BunPlugin = {
  name: "Plaited loader",
  setup(build) {
    build.onResolve({ filter: /\/route.tsx?$/ }, async ({ path }) => {
      // mark external all non-template files
    })
    build.onLoad({ filter: /\/route.tsx?$/ }, async ({ path }) => {
      // modify the contents of the file to be an export object.
      return {
        contents,
        loader: "tsx",
      };
    });
  },
}