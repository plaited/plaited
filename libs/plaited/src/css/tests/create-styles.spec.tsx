import { test, expect } from 'bun:test'
import { createStyles } from '../create-styles.js'
import { assignStyles } from '../assign-styles.js'
test('createStyles: simple rules', () => {
  const s = createStyles({
    base: {
      fontzSize: `16px`,
      lineHeight: 1.5,
      color: 'rgb(60,60,60)',
    },
    highlighted: {
      color: 'rebeccapurple',
    },
  })
  expect(s).toMatchSnapshot()
})

test('createStyles: Pseudo-classes', () => {
  expect(
    createStyles({
      button: {
        backgroundColor: 'lightblue',
      },
    }),
  ).toMatchSnapshot()
  expect(
    createStyles({
      button: {
        backgroundColor: {
          default: 'lightblue',
          ':hover': 'blue',
          ':active': 'darkblue',
        },
      },
    }),
  ).toMatchSnapshot()
})

test('createStyles: Pseudo-elements', () => {
  expect(
    createStyles({
      input: {
        // pseudo-element
        '::placeholder': {
          color: '#999',
        },
        color: {
          default: '#333',
          // pseudo-class
          ':invalid': 'red',
        },
      },
    }),
  ).toMatchSnapshot()
})

test('createStyles: media query', () => {
  expect(
    createStyles({
      base: {
        width: {
          default: 800,
          '@media (max-width: 800px)': '100%',
          '@media (min-width: 1540px)': 1366,
        },
      },
    }),
  ).toMatchSnapshot()
})

test('createStyles: combining conditions', () => {
  expect(
    createStyles({
      button: {
        color: {
          default: 'var(--blue-link)',
          ':hover': {
            default: null,
            '@media (hover: hover)': 'scale(1.1)',
          },
          ':active': 'scale(0.9)',
        },
      },
    }),
  ).toMatchSnapshot()
})

test('JSX: spread', () => {
  const s = createStyles({
    button: {
      color: {
        default: 'var(--blue-link)',
        ':hover': {
          default: null,
          '@media (hover: hover)': 'scale(1.1)',
        },
        ':active': 'scale(0.9)',
      },
    },
  })
  expect(<button {...s.button}></button>).toMatchSnapshot()
})
