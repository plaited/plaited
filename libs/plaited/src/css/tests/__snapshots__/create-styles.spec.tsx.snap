// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`createStyles: simple rules 1`] = `
{
  "base": {
    "className": "base__1f8ei0o p_2h8kl40 p4c01gd6 p10v3vx6",
    "stylesheet": [
      ".p_2h8kl40{fontz-size:16px;}",
      ".p4c01gd6{line-height:1.5;}",
      ".p10v3vx6{color:rgb(60,60,60);}",
    ],
  },
  "highlighted": {
    "className": "highlighted_g28n4j p91m869",
    "stylesheet": [
      ".p91m869{color:rebeccapurple;}",
    ],
  },
}
`;

exports[`createStyles: Pseudo-classes 1`] = `
{
  "button": {
    "className": "button__8iz41n p2l5cglg",
    "stylesheet": [
      ".p2l5cglg{background-color:lightblue;}",
    ],
  },
}
`;

exports[`createStyles: Pseudo-classes 2`] = `
{
  "button": {
    "className": "button_1tjvtxt p2l5cglg p1sj0k0a p3ni2mzo",
    "stylesheet": [
      ".p2l5cglg{background-color:lightblue;}",
      ".p1sj0k0a{&:hover{background-color:blue;}}",
      ".p3ni2mzo{&:active{background-color:darkblue;}}",
    ],
  },
}
`;

exports[`createStyles: Pseudo-elements 1`] = `
{
  "input": {
    "className": "input_1etkoss pqpi8rk p_1uwdrog",
    "stylesheet": [
      ".pqpi8rk{color:#333;}",
      ".p_1uwdrog{&:invalid{color:red;}}",
    ],
  },
}
`;

exports[`createStyles: media query 1`] = `
{
  "base": {
    "className": "base_3d4s76s paqr07x p_3zvevs7 p_l1op19",
    "stylesheet": [
      ".paqr07x{width:800;}",
      ".p_3zvevs7{@media (max-width: 800px){width:100%;}}",
      ".p_l1op19{@media (min-width: 1540px){width:1366;}}",
    ],
  },
}
`;

exports[`createStyles: combining conditions 1`] = `
{
  "button": {
    "className": "button_mb2ltj p_2dr9p51 p_d33wg2 p_1bahz5i",
    "stylesheet": [
      ".p_2dr9p51{color:var(--blue-link);}",
      ".p_d33wg2{&:hover{@media (hover: hover){color:scale(1.1);}}}",
      ".p_1bahz5i{&:active{color:scale(0.9);}}",
    ],
  },
}
`;

exports[`JSX: spread 1`] = `
{
  "$": "ðŸ¦„",
  "client": [
    "<button ",
    "class="button_mb2ltj p_2dr9p51 p_d33wg2 p_1bahz5i" ",
    ">",
    "</button>",
  ],
  "registry":   Set {},
  "server": [
    "<button ",
    "class="button_mb2ltj p_2dr9p51 p_d33wg2 p_1bahz5i" ",
    ">",
    "</button>",
  ],
  "stylesheets":   
Set {
    ".p_2dr9p51{color:var(--blue-link);}",
    ".p_d33wg2{&:hover{@media (hover: hover){color:scale(1.1);}}}",
    ".p_1bahz5i{&:active{color:scale(0.9);}}",
  }
,
}
`;
